src/lib/Libtpp/tpp_gss.c

```diff
@@ -0,0 +1,588 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#include <pbs_config.h>   /* the master config generated by configure */
+
+#if defined(PBS_SECURITY) && (PBS_SECURITY == KRB5)
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <gssapi.h>
+
+#include "rpp.h"
+#include "tpp_common.h"
+#include "libutil.h"
+
+#include "pbs_gss.h"
+
+pthread_mutex_t gss_lock;
+
+/* the function pointer to the upper layer received packet handler */
+int (*tpp_pkt_handler)(int tfd, void *data, int len, void *ctx, void *extra);
+
+/* the function pointer to the upper layer connection close handler */
+int (*tpp_close_handler)(int tfd, int error, void *ctx, void *extra);
+
+/* the function pointer to the upper layer connection restore handler */
+int (*tpp_post_connect_handler)(int tfd, void *data, void *ctx, void *extra);
+
+/* the function pointer to the upper layer pre packet send handler */
+int (*tpp_pkt_presend_handler)(int tfd, tpp_packet_t *pkt, void *extra);
+
+/* the function pointer to the upper layer post packet send handler */
+int (*tpp_pkt_postsend_handler)(int tfd, tpp_packet_t *pkt, void *extra);
+
+/* upper layer timer handler */
+int (*tpp_timer_handler)(time_t now);
+
+/**
+ * @brief
+ * 	-tpp_gss_logerror - logs the provided error
+ *
+ * @param[in] func_name - function name that invokes the error
+ * @param[in] msg - error message
+ */
+void
+tpp_gss_logerror(const char *func_name, const char* msg)
+{
+	tpp_log_func(LOG_ERR, func_name, (char*) msg);
+}
+
+/**
+ * @brief
+ * 	-tpp_gss_logdebug - logs the provided debug message
+ *
+ * @param[in] func_name - function name that invokes the message
+ * @param[in] msg - error message
+ */
+void
+tpp_gss_logdebug(const char *func_name, const char* msg)
+{
+	tpp_log_func(LOG_DEBUG, func_name, (char*) msg);
+}
+
+/**
+ * @brief
+ * 	-tpp_gss_set_extra_host - allocates the hostname to gss_extra structure
+ *
+ * @param[in] extra - gss extra structure as void*
+ * @param[in] hostname - hostname
+ *
+ * @return	int
+ * @retval	PBS_GSS_OK on success
+ * @retval	!= PBS_GSS_OK on error
+ */
+int
+tpp_gss_set_extra_host(void *extra, char *hostname)
+{
+	pbs_gss_extra_t *gss_extra = (pbs_gss_extra_t*) extra;
+	char *hn;
+
+	if (gss_extra->hostname)
```

Comment:
shouldn't we check if hostname == gss_extra->hostname?

Reply:
No, the gss_extra->hostname is supposed to be always set to the GSS server hostname. No matter we are the GSS server or GSS client.

* For GSS client: The parameter 'hostname' conveys the GSS server hostname and it is used.
* For GSS server: The parameter 'hostname' is not used because it is set to the GSS client hostname, so we do not want to use it. The server name is retrieved by gethostname() in this case.

I will improve the comment of the function tpp_gss_set_extra_host().